File: leetcode/0622.设计循环队列/Solution.go
type MyCircularQueue struct {
	buffer []int
	front  int
	rear   int
}

func Constructor(k int) MyCircularQueue {
	return MyCircularQueue{
		buffer: make([]int, k+1),
		front:  0,
		rear:   0,
	}
}

func (this *MyCircularQueue) EnQueue(value int) bool {
	if this.IsFull() {
		return false
	}
	this.buffer[this.rear] = value
	this.rear = (this.rear + 1) % len(this.buffer)
	return true
}

func (this *MyCircularQueue) DeQueue() bool {
	if this.IsEmpty() {
		return false
	}
	this.front = (this.front + 1) % len(this.buffer)
	return true
}

func (this *MyCircularQueue) Front() int {
	if this.IsEmpty() {
		return -1
	}
	return this.buffer[this.front]
}

func (this *MyCircularQueue) Rear() int {
	if this.IsEmpty() {
		return -1
	}
	return this.buffer[(this.rear-1+len(this.buffer))%len(this.buffer)]
}

func (this *MyCircularQueue) IsEmpty() bool {
	return this.front == this.rear
}

func (this *MyCircularQueue) IsFull() bool {
	return (this.rear+1)%len(this.buffer) == this.front
}

File: leetcode/0622.设计循环队列/Solution.ts
class MyCircularQueue {
  private buffer: number[]
  private pHead: number
  private pTail: number

  private equalCause: 'full' | 'empty'

  constructor(k: number) {
    this.buffer = new Array(k)
    this.pHead = 0
    this.pTail = 0
    this.equalCause = 'empty'
  }

  enQueue(value: number): boolean {
    if (this.isFull()) return false
    this.buffer[this.pHead++] = value
    this.pHead %= this.buffer.length
    console.log('pHead', this.pHead)
    if (this.pHead === this.pTail) {
      this.equalCause = 'full'
    }
    return true
  }

  deQueue(): boolean {
    if (this.isEmpty()) return false
    this.pTail++
    this.pTail %= this.buffer.length
    if (this.pTail === this.pHead) {
      this.equalCause = 'empty'
    }
    return true
  }

  Front(): number {
    if (this.isEmpty()) return -1
    return this.buffer[this.pTail]
  }

  Rear(): number {
    if (this.isEmpty()) return -1
    // console.log(this.buffer)
    return this.buffer[(this.pHead - 1 + this.buffer.length) % this.buffer.length]
  }

  isEmpty(): boolean {
    return this.pHead === this.pTail && this.equalCause === 'empty'
  }

  isFull(): boolean {
    return this.pHead === this.pTail && this.equalCause === 'full'
  }
}
